src/as400_sha_fmt_plug.c

```diff
@@ -0,0 +1,419 @@
+    while(strlen(username)<AS400_USERID_LENGTH)
+    {
+        strcat(username," ");
+    }		
+	
+    // Store username in first 10 characters of return value so that we can access username in get_salt
+	strcpy(retval,username);
+	// Concatenate hash from inputfile to return value
+	strcat(retval,fields[1]);
+
+	return(retval);
+}
+
+
+// ==================================================================================================
+// get_salt fills the salt structure with data for later use
+
+static void *get_salt(char *ciphertext)
+{
+	int i,len;
+	static struct custom_salt cs;
+	
+	// Temporary variables to hold username and hash
+	char username[AS400_USERID_LENGTH+1];
+	char hash[CIPHERTEXT_LENGTH+1];
+	
+	// Split ciphertext into username and hash (because these were concatenated in prepare())
+	for(i = 0; i < AS400_USERID_LENGTH; i++)
+		username[i] = ciphertext[i];
+	username[AS400_USERID_LENGTH] = '\0';
+	
+	for(i = 0; i < CIPHERTEXT_LENGTH; i++)
+		hash[i] = ciphertext[i+AS400_USERID_LENGTH];
+	hash[i] = '\0';
+	
+	// convert ciphertext to binary and store for later use 
+	len = strlen(hash) >> 1;
+	for (i = 0; i < len; i++)
+		cs.storedbinaryhash[i] = (atoi16[ARCH_INDEX(hash[i << 1])] << 4) + atoi16[ARCH_INDEX(hash[(i << 1) + 1])];
+
+	// Store userid as utf-16be for later use (we do this here for performance reasons)
+	for(i=0;i<AS400_USERID_LENGTH;i++)
+	{
+		cs.username_utf16be[(i*2)]=0x00;
+		cs.username_utf16be[(i*2)+1]=username[i];
+	}
+	
+	return (void*)&cs;
+}
+
+
+// ==================================================================================================
+// Retrieve the current salt from memory for use in crypt_all() function
+
+static void set_salt(void *salt)
+{
+	cur_salt = (struct custom_salt *)salt;
+}
+
+
+// ==================================================================================================
+// Store the password candidates that are passed from the JtR engine
+// in our own variables for later use. 
+
+static void as400_set_key(char *key,int index)
+{
+	strnzcpy(as400_saved_passwords[index],key,strlen(key)+1);
+
+}
+
+
+// ==================================================================================================
+// Return the passwords candidate that is stored at 'index'
+// This function is used by JtR engine to retrieve the plaintext password associated with a cracked hash 
+
+static char *as400_get_key(int index)
+{
+	return as400_saved_passwords[index];
+}
+
+
+// ==================================================================================================
+// function to compute AS/400 hash
+
+static void as400_password_hash(const char *userid_utf16be, const char *password, uint8_t *hash)
+{ 
+    int i;
+	char password_utf16be[PLAINTEXT_LENGTH*2];
+	
+	SHA_CTX s_ctx;
+	uint8_t digest[SHA_DIGEST_LENGTH];
+
+	// convert password to utf-16be
+	for(i=0;i<(strlen(password));i++)
+	{
+		password_utf16be[(i*2)]=0x00;
+		password_utf16be[(i*2)+1]=password[i];
+	}
```

Comment:
This is not a proper UTF-16 conversion unless input is ISO-8859-1. We have functions that do it better, see eg. oracle_fmt_plug.c